pulp/apis/cuopt_api.py

```diff
@@ -0,0 +1,337 @@
+            self.model = data_model.DataModel()
+            self.settings = solver_settings.SolverSettings()
+            self.solver = solver
+            self.solution = None
+            self.var_list = None
+            self.solver_params = solverParams
+
+            ## TODO: Disable logging if self.msg = False
+            for key, value in solverParams.items():
+                if key == "optimality_tolerance":
+                    self.settings.set_optimality_tolerance(value)
+
+
+        def findSolutionValues(self, lp):
+            model = lp.solverModel
+            solution = self.solution
+            solutionStatus = solution.get_termination_reason()
+            print("CUOPT status=", solutionStatus)
+
+            CuoptLpStatus = {
+                0: LpStatusNotSolved,
+                1: LpStatusOptimal,
+                2: LpStatusInfeasible,
+                3: LpStatusUnbounded,
+                4: LpStatusNotSolved,
+                5: LpStatusNotSolved,
+                6: LpStatusNotSolved, # Primal Feasible?
+            }
+
+            CuoptMipStatus = {
+                0: LpStatusNotSolved,
+                1: LpStatusOptimal,
+                2: LpStatusNotSolved, # Feasible Solution Found ??
+                3: LpStatusInfeasible,
+                4: LpStatusUnbounded,
+            }
+
+            if self.msg:
+                print("CUOPT status=", solutionStatus)
+
+            lp.resolveOK = True
+            for var in lp._variables:
+                var.isModified = False
+
+            if solution.get_problem_category() == 0:
+                status = CuoptLpStatus.get(solutionStatus, LpStatusUndefined)
+            else:
+                status = CuoptMipStatus.get(solutionStatus, LpStatusUndefined)
+
+            lp.assignStatus(status)
+
+            values = solution.get_primal_solution()
+
+            for var, value in zip(lp._variables, values):
+                var.varValue = value
+
+            if not solution.get_problem_category():
+                # TODO: Compute Slack
+
+                redcosts = solution.get_lp_stats()["reduced_cost"]
+                for var, value in zip(lp._variables, redcosts):
+                    var.dj = value
+
+                duals = solution.get_dual_solution()
+                for constr, value in zip(lp.constraints.values(), duals):
+                    constr.pi = value
+
+            return status
+
+        def available(self):
+            """True if the solver is available"""
+            return True
+
+        def callSolver(self, lp, callback=None):
+            """Solves the problem with CUOPT"""
+            self.solveTime = -clock()
+            # TODO: Add callback
+            log_file = self.optionsDict.get("logPath") or ""
+
+            self.settings.set_infeasibility_detection(True)
+            self.solution = self.solver.Solve(lp.solverModel, self.settings, log_file)
```

Comment:
most other solvers store the solution inside `lp.solverModel`, I see this is not the case for cuopt by default.

Instead of creating an attribute  `self.solution` to store it in the solver, I think it's cleaner if callSolver returns the content of `self.solver.Solve` and that gets passed to `findSolutionValues` as an additional argument named `solution`. 

Another option is to attach the `solution` attribute to `lp.solverModel` and then you get everything from the object `lp`.

Reply:
Added solution as return object from callSolver. Refer #839